### ISSUE TITLE
Join us as a new Taichi contributor!

### ISSUE THREAD

**Issue by lucywsq** [taichi-dev/taichi#3151]
Below are features that are relatively friendly to new comers. Everyone is welcome to participate :-)
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-403e542b-7fff-8acf-11d2-f6591ba3c26f"><div dir="ltr" style="margin-left:0pt;" align="left">

**Welcome contribution**

Issue | Link
-- | --
Support more atomic operations | https://github.com/taichi-dev/taichi/issues/2675
Support efficient operations for large matrix | https://github.com/taichi-dev/taichi/issues/2696
[refactor] Make Callable::Args/Ret serializable | https://github.com/taichi-dev/taichi/issues/2625
Deprecate Expr::operator= | https://github.com/taichi-dev/taichi/issues/2684
[vulkan] Clear unnecessary GLSL shader files on Vulkan backend | https://github.com/taichi-dev/taichi/issues/2711
Put repeating code blocks in Github CI workflows into scripts. | https://github.com/taichi-dev/taichi/issues/2715
Invoke materialize() only before calling a Taichi kernel | https://github.com/taichi-dev/taichi/issues/2730

</div></b>

**Comment by lucywsq** [comment]
These issues haven't been assigned yet. You can still choose one to do some excellent work : )

Put repeating code blocks in Github CI workflows into scripts. | #2715
-- | --
[lang] Expose thread index in @ti.kernel | #2725
Support more atomic operations | #2675
Support efficient operations for large matrix | #2696
Deprecate Expr::operator= | #2684

**Comment by hantsy** [comment]
Why not add these issues to the yearly hacker event: https://hacktoberfest.digitalocean.com/, and hackerfest hashtag on Twitter.

**Comment by lucywsq** [comment]
@hantsy Thanks for your suggestion :)

**Comment by lucywsq** [comment]
Here are some new issues with the tag "good first issue". Welcome to take one and implement it. 
We are looking forward to your contributions! : )


Issue | Link
-- | --
[refactor] Clean up keep_raw | https://github.com/taichi-dev/taichi/issues/2718
Support setting the random seed during initialization | https://github.com/taichi-dev/taichi/issues/2734
Exception raising in Field APIs | https://github.com/taichi-dev/taichi/issues/2846
[docs] Add doc about building python part without compiling taichi-core | https://github.com/taichi-dev/taichi/issues/3056

**Comment by lucywsq** [comment]
Currently, if you run `pylint python/taichi/`there're a lot of warnings/errors.
We'd like to get rid of these and run pylint in CI. You're welcome to submit one PR per pylint checker/rule.
Most of these rules are easy to fix, so please feel free to give them a try!
If you are interested in fixing one of these rules, please leave a message here to let us know! Thanks! 

Besides，find out all errors for one rule: 

 `pylint python/taichi/ --disable=all --enable=C0XXX # replace this with your rule number`

Submit a PR to taichi: 
please make sure the rule is enabled in CI along with your PR by simply adding the rule to the end of the list here. https://github.com/taichi-dev/taichi/blob/master/.github/workflows/presubmit.yml#L91 

目前 Taichi 中输入 `pylint python/taichi/` 命令后，会出现比较多的报错信息。
未来我们希望在 CI 中成功运行  pylint ,  让体验更为顺滑。在此邀请广大开发者们一起助力，共同维护 pylint 提示，选择其中某个类别的命令完成修复，通过提 PR 的方式进行贡献。
特别提示：这些 pylint 命令都比较简单，所以不用担心难度，放心大胆的来练练手！
参与方式：如果对修复某个命令感兴趣，请在本帖下留言，我们将在 assignee 这栏中留下你的名字哦。
感谢每一位开发者的支持！

补充：这个命令行可以看到所有的报错信息：

 `pylint python/taichi/ --disable=all --enable=C0XXX # replace this with your rule number`

同时，请确保修好的rule 要在CI里面打开相应的pylint检查，确保CI能够正常通过哦～


规则名称 | 累计代码行数 | 命令名称 | Assignee
-- | -- | -- | --
W0621 | 57 | redefined-outer-name |  
W0622 | 53 | redefined-builtin |  
W0401 | 2 | wildcard-import |  
C0209 | 67 | consider-using-f-string |  
W0404 | 19 | reimported |  
W0611 | 45 | unused-import |  
W0612 | 134 | unused-variable |  
E1101 | 72 | no-member |  
R0402 | 2 | consider-using-from-import |  
R0201 | 14 | no-self-use |  
W0235 | 3 | useless-super-delegation |  
R1705 | 73 | no-else-return |  
C0200 | 17 | consider-using-enumerate |  
R0205 | 5 | useless-object-inheritance |  
R1732 | 2 | consider-using-with |  
W0102 | 30 | dangerous-default-value |  
W0101 | 8 | unreachable |  
R1710 | 19 | inconsistent-return-statements |  
R1703 | 2 | simplifiable-if-statement |  
W0108 | 18 | unnecessary-lambda |

**Comment by jyz0309** [comment]
I want to pick up the W0401, Can you assign it to me?@lucywsq

**Comment by LJHG** [comment]
I want to pick up the R1703, Can you assign it to me?@lucywsq

**Comment by Dustyposa** [comment]
I want to pick up the W0201, Can you assign it to me?@lucywsq

**Comment by tczhangzhi** [comment]
I want to pick up the W0611, Can you assign it to me?@lucywsq

**Comment by Dustyposa** [comment]
I want to pick up the W0404, Can you assign it to me?@lucywsq

**Comment by kemingy** [comment]
I want to pick up the C0200, Can you assign it to me? @lucywsq

**Comment by lucywsq** [comment]
@Dustyposa And you may skip this rule W0102(it seems complex). Thanks!

**Comment by zstone12** [comment]
I want to pick up the C0209, Can you assign it to me? @lucywsq

**Comment by Dustyposa** [comment]
> And you may skip this rule W0102(it seems complex). Thanks!

OK. It looks need a detailed description for it. Can solve the question.

**Comment by nasnoisaac** [comment]
I want to pick up the W0108, Can you assign it to me? @lucywsq

**Comment by imdouyu** [comment]
I want to pick up the R0402, Can you assign it to me? @lucywsq

**Comment by MerleLiuKun** [comment]
I want to pick up the W0612, Can you assign it to me? @lucywsq :)

**Comment by IceCodeNew** [comment]
I would like to take care of `C0209`, can you assign it to me?
@lucywsq

**Comment by k-ye** [comment]
@imdouyu @MerleLiuKun @IceCodeNew  thank you all for expressing the interests! Assigned :-)

**Comment by SFAfreshman** [comment]
i want try to pickup` W0101`,can you assign it to me? @lucywsq

**Comment by darkSheep404** [comment]
I want to pick up the W0235, Can you assign it to me?@lucywsq

**Comment by skyzh** [comment]
I'll pick R0201.

**Comment by sjwsl** [comment]
Hi contributors, thanks for your enthusiasm and great works! After submitting a PR, you may see

<img width="933" alt="Screenshot 2021-11-13 at 21 04 44" src="https://user-images.githubusercontent.com/44310559/141644834-8fe69e10-0518-4520-a307-0e69d1abc522.png">

That's because some new changes in `master` branch have conflicts with yours (most likely in `.github/workflows/presubmit.yml`). Please click the button `Resolve conflicts` to unblock the merge. If unfortunately the button is forbidden, please manually fix it by

```bash
git remote add upstream https://github.com/taichi-dev/taichi.git
git pull upstream master
```

Then you need to resolve the conflicts and push the new commit. 

```bash
# Edit the files containing conflicts
git add .
git commit
git push
```

We're sorry since there are frequent new changes into the master branch, you may need to resolve conflicts and wait for CI more than once. Thanks so much for your patience, your help will greatly speed up our process of PRs.

Again, Thanks for your contributions and welcome to our community!

Cheers :)

**Comment by licktion** [comment]
R1710 plz ^_^

**Comment by bzd111** [comment]
I'll pick W0235.

**Comment by Lyken17** [comment]
I'll pick W0101 and here is PR to add the lint.

**Comment by PragmaTwice** [comment]
W0108 plz 🍰

**Comment by MarkHershey** [comment]
I'll take W0622~

**Comment by DeepDuke** [comment]
I want to pick up the E1101, Can you assign it to me? @lucywsq

**Comment by sjwsl** [comment]
Hi contributors! @MarkHershey @DeepDuke @bzd111 @licktion @MerleLiuKun @Lyken17 @ZonePG

We are sorry that you had to resolve conflicts with `master` again and again. The good news is that after #3503 we have put each rule on a separate line so that your PRs should have no conflicts with others. Now please kindly make the final step before we can merge it

1. Overwrite your previous changes in `presubmit.yml` with `master`'s in web or command line.
https://github.com/taichi-dev/taichi/blob/18b947206b18028b484537a426fa5cd84e3eb8d8/.github/workflows/presubmit.yml#L91
2. Mark your new pylint rule as `True` in `scripts/generate_pylint_tags.py`.
3. Push your changes and see if the conflicts disappear.

Everything is fine now! You should no longer need to resolve conflicts in `presubmit.yml`. However, there may still be non-trivial conflicts in Python scripts and of course, you need to resolve them. Feel free to reach us if you still have difficulty making it work.

Hope you enjoyed it!

**Comment by Lyken17** [comment]
@sjwsl Sure, rule W0101  fixed. Please take a time and have a check https://github.com/taichi-dev/taichi/pull/3497

**Comment by DeepDuke** [comment]
@sjwsl Conflicts resolved. Please take a time and have a check #3500

**Comment by ZonePG** [comment]
@sjwsl Conflicts resolved. Please take a time and have a check #3498

**Comment by lucywsq** [comment]
For more discussion, please join this[ slack channel ](https://join.slack.com/share/enQtMjc0MjY3MzMzNzY1Mi1kNTllMWY5Yjk4ZDM4YmU5MjQ2MzM5MmIzZDUzZmI2MGM4MGVmMjc2ZTU5Y2RjMmVhMDhjZTkxMzY1MDJmODk0), then we would respond to your problems on time. Thanks for all the great work!!

**Comment by skyzh** [comment]
Hi @lucywsq , the Slack link you sent is a channel link and is not supported by community edition of Slack.

> Slack Connect makes working with people from other companies as fast and simple as working with your teammates. Collaborate in channels, share files, and send direct messages — all with Slack’s enterprise-grade security. 

You may invite us directly to Taichi's workspace.

**Comment by skyzh** [comment]
> For more discussion, please join this[ slack channel ](https://join.slack.com/share/enQtMjc0MjY3MzMzNzY1Mi1kNTllMWY5Yjk4ZDM4YmU5MjQ2MzM5MmIzZDUzZmI2MGM4MGVmMjc2ZTU5Y2RjMmVhMDhjZTkxMzY1MDJmODk0) then we would respond to your problems on time. Thanks for all the great work!!

So to join the channel, we will need to first create our own workspace, and then join the channel as another workspace's member. This doesn't sound like a common practice for open-source commnities.

**Comment by skyzh** [comment]
... and we only have 90-days trial for joining.

<img width="595" alt="image" src="https://user-images.githubusercontent.com/4198311/142348583-c7c179df-b0df-44d0-b2db-d84f2999b0ca.png">

**Comment by lucywsq** [comment]
@skyzh sorry for that. We'll find better ways for communicating.

**Comment by DeepDuke** [comment]
<img width="746" alt="Screen Shot 2021-11-19 at 12 10 22 AM" src="https://user-images.githubusercontent.com/15854511/142452992-a6dd6774-c203-4813-b3ba-93162c0223f9.png">
cannot join now ...

**Comment by sjwsl** [comment]
Hi all! Please join our Slack workspace for discussion via [this invitation link](https://join.slack.com/t/taichicommunity/shared_invite/zt-yz96moub-18KdjyCfsR8i6q3SirghfQ). We apologize for the inconvenience caused.
Thanks!

**Comment by lucywsq** [comment]
Before that, please prepare one Gmail account or an Apple account.
If that link has broken, please send an email to siqiwu@taichi.graphics, then we will send another invitation.

**Comment by lucywsq** [comment]
Hi all, we are so appreciative of your contribution to fixing pylint rules through the “Let’s become Taichi contributors within 10 minutes” program. If you have finished this task, please send an email to **siqiwu@taichi.graphics** with your offline address. We would deliver a hoodie to you. : ) Looking forward to hearing from you!

**Comment by kxxt** [comment]
I have fixed W1309, #3757 .
